### DISCUSSION TITLE
Integer enum query parameters validation always fails

### DISCUSSION THREAD

**Discussion by AlexKLWS** [fastapi/fastapi#7133] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from enum import Enum
from fastapi import FastAPI

app = FastAPI()


class Category(int, Enum):
    MISC = 0
    LIFE = 1
    CODE = 2
    GUIDES = 3
    PROJECTS = 4


@app.get("/stuff")
def read_root(category: Category):
    return {"Hello": "World"}
```

Run `$uvicorn test:app`

### Description

* Open the browser and call the endpoint `/stuff?category=1`.
* It returns a 422 error `"value is not a valid enumeration member; permitted: 0, 1, 2, 3, 4"` even though the correct int enum value is provided. 
* **Expected:** It returns a JSON with `{"Hello": "World"}`.

### Environment

* OS: macOS
* FastAPI Version: 0.61.1
* Python version: 3.8.5

### Additional context

I've looked through all pydantic issues, found [this one](https://github.com/samuelcolvin/pydantic/issues/1735), but I don't know if it's related, or what it's current status (whether it has been resolved).

**Comment by AlexKLWS** [comment]
With `IntEnum` type validation works as expected

**Comment by iblasi** [comment]
@AlexKLWS thank you for your answer to solve the issue.
Just for other people as me that found this issue interesting, i will add the code required to avoid misunderstanding of where to put ```IntEnum```.

```
from enum import Enum, IntEnum
from fastapi import FastAPI

app = FastAPI()

class Category(IntEnum, Enum):
    MISC = 0
    LIFE = 1
    CODE = 2
    GUIDES = 3
    PROJECTS = 4

@app.get("/stuff")
def read_root(category: Category):
    return {"Hello": "World"}
```

**Comment by Kludex** [comment]
@iblasi `IntEnum` is already an `Enum` :kissing: 

Cleaner:
```python
class Category(IntEnum):
    MISC = 0
    LIFE = 1
    CODE = 2
    GUIDES = 3
    PROJECTS = 4
```

Reference: https://github.com/python/cpython/blob/3.9/Lib/enum.py#L798

**Comment by connebs** [comment]
I've actually noticed an issue with this: in the OpenAPI schema, the name of each enum doesn't show up, only the values `(0, 1, 2, 3, 4)`.

Is it possible to show the names in the schema? The whole point of an `IntEnum` is to put names to integers, but consumers of your API can only see the values not the (very helpful) names.

**Comment by Kludex** [comment]
You sure that's the point? 🤔
Can you provide me a real case scenario?

**Comment by connebs** [comment]
Sure, using the IntEnum example as above, all you see is this:

```
category: [0, 1, 2, 3, 4]
```

But as a consumer of the API, I have no idea what those numbers mean. I need to see their human-readable names in order to understand my options. If the names were listed in the schema, I would know that:

```python
MISC = 0
LIFE = 1
CODE = 2
GUIDES = 3
PROJECTS = 4
```

**Comment by est** [comment]
had the same issue with query parameters. 

`status: Optional[int] = Query(None, enum=Enum("status", {"enabled": 1, "disabled": 0}))`

this doesn't work.

**Comment by ViieeS** [comment]
@acnebs It's the problem I faced with recently. I generate models and services for UI client from openapi using openapi-generator-cli and enums looks like:

**Python:**

```py
class MyEnum(IntEnum):
    Foo = 1
    Bar = 2
    Baz = 3
```

**TypeScript**
```ts
export enum MyEnum {
    NUMBER_1 = 1,
    NUMBER_2 = 2,
    NUMBER_3 = 3
}
```

Did you find a solution?

**Comment by oligond** [comment]
@acnebs To have the OpenAPI docs display the names instead of or additionally to the integer values of your enum, you could try tweaking the corresponding JSON schema:
```python
class Category(IntEnum):
    MISC = 0
    LIFE = 1
    CODE = 2
    GUIDES = 3
    PROJECTS = 4

    @classmethod
    def __modify_schema__(cls, schema: Dict[str, Any]):
        schema["enum"] = [f"{choice.name} ({choice.value})" for choice in cls]
        return schema
```
This sample code should result in your OpenAPI docs containing entries such as `MISC (0)` for each enum member.

If you want to generalize this, let's say for all enums used in your models, you can subclass `IntEnum`:

```python
from enum import IntEnum

class WellDocumentedIntEnum(IntEnum):

    @classmethod
    def __modify_schema__(cls, schema: Dict[str, Any]):
        schema["enum"] = [f"{choice.name} ({choice.value})" for choice in cls]
        return schema


class Category(WellDocumentedIntEnum):
    MISC = 0
    LIFE = 1
    CODE = 2
    GUIDES = 3
    PROJECTS = 4
```

**Reply by arisliang** [reply]
This does not seem to work anymore, is it?

>pydantic.errors.PydanticUserError: The `__modify_schema__` method is not supported in Pydantic v2. Use `__get_pydantic_json_schema__` instead in class 'EnumXXX'.

**Comment by connebs** [comment]
That's a great solution, thanks @oligond.

**Comment by ViieeS** [comment]
Thanks @oligond for the general idea! Solution for  [typescript-axios](https://openapi-generator.tech/docs/generators/typescript-axios/):
```py
field_schema["x-enum-varnames"] = [choice.name for choice in cls]
```
Also you don't have to return anything:

> `__modify_schema__` should mutate the dict it receives in place, the returned value will be ignored

https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__

**Comment by connebs** [comment]
Just actually tried this out, and the slight problem with this solution (and presumably why it wasn't done before) is that interfaces which consume the schema (e.g. SwaggerUI) assume these are the actual values and thus use them directly in the "Example Value".

e.g. SwaggerUI shows this as an example of how to use the endpoint:

```json
{
  "category": "LIFE (1)"
}
```

When of course it should actually be showing:

```json
{
  "category": 1
}
```

**Comment by invokermain** [comment]
@ViieeS really useful thanks should somehow be a lot more visible!

**Comment by tiangolo** [comment]
Thanks for the conversation everyone! And thanks @AlexKLWS for coming back to close the issue. ☕

**Comment by auxsvr** [comment, marked as answer]
The code in https://github.com/tiangolo/fastapi/discussions/7133#discussioncomment-5134570 updated to Pydantic 2 and OpenAPI 3.1:
```python
class OpenAPIIntEnum(IntEnum):
    """Document names and values for IntEnum."""

    def __new__(cls, value, doc=None):
        """Add docstrings to attributes."""
        self = int.__new__(cls, value)
        self._value_ = value
        if doc is not None:
            self.__doc__ = doc
        return self

    @classmethod
    def __get_pydantic_json_schema__(cls, core_schema: CoreSchema, handler: GetJsonSchemaHandler):
        json_schema = BaseModel.__get_pydantic_json_schema__(core_schema, handler)
        json_schema["x-enum-varnames"] = [v.name for v in cls]  # Does not work with Swagger.
        json_schema["oneOf"] = [
            {"title": v.name, "const": v.value, "description": inspect.getdoc(v)} for v in cls
        ]  # For OpenAPI 3.1
        json_schema = handler.resolve_ref_schema(json_schema)
        return json_schema
```

**Reply by ttlns** [reply]
adding
```python
if "enum" in json_schema: del json_schema["enum"]
```
before the return to avoid the duplicate description


Before
<img width="680" height="428" alt="image" src="https://github.com/user-attachments/assets/e5624fe2-0a1c-4bec-9a6e-598e2e5fad9c" />
